Spigot-API-Patches/0006-Adventure-Events.patch

```diff
@@ -0,0 +1,1062 @@
+     private Boolean cancel;
+ 
++    @Deprecated // Paper
+     public PlayerKickEvent(@NotNull final Player playerKicked, @NotNull final String kickReason, @NotNull final String leaveMessage) {
++        super(playerKicked);
++        this.kickReason = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(kickReason); // Paper
++        this.leaveMessage = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserialize(leaveMessage); // Paper
++        this.cancel = false;
++    }
++    // Paper start
++    public PlayerKickEvent(@NotNull final Player playerKicked, @NotNull final net.kyori.adventure.text.Component kickReason, @NotNull final net.kyori.adventure.text.Component leaveMessage) {
+         super(playerKicked);
+         this.kickReason = kickReason;
+         this.leaveMessage = leaveMessage;
+         this.cancel = false;
+     }
+ 
++    /**
++     * Gets the leave message send to all online players
++     *
++     * @return string kick reason
++     */
++    public @org.jetbrains.annotations.Nullable net.kyori.adventure.text.Component leaveMessage() {
++        return this.leaveMessage;
++    }
++
++    /**
++     * Sets the leave message send to all online players
++     *
++     * @param leaveMessage leave message
++     */
++    public void leaveMessage(@org.jetbrains.annotations.Nullable net.kyori.adventure.text.Component leaveMessage) {
++        this.leaveMessage = leaveMessage;
++    }
++
+     /**
+      * Gets the reason why the player is getting kicked
+      *
+      * @return string kick reason
+      */
++    public @org.jetbrains.annotations.Nullable net.kyori.adventure.text.Component reason() {
++        return this.kickReason;
++    }
++
++    /**
++     * Sets the reason why the player is getting kicked
++     *
++     * @param kickReason kick reason
++     */
++    public void reason(@org.jetbrains.annotations.Nullable net.kyori.adventure.text.Component kickReason) {
++        this.kickReason = kickReason;
++    }
++    // Paper end
++
++    /**
++     * Gets the reason why the player is getting kicked
++     *
++     * @return string kick reason
++     * @deprecated in favour of {@link #reason()}
++     */
+     @NotNull
++    @Deprecated // Paper
+     public String getReason() {
+-        return kickReason;
++        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().serialize(this.kickReason); // Paper
+     }
+ 
+     /**
+      * Gets the leave message send to all online players
+      *
+      * @return string kick reason
++     * @deprecated in favour of {@link #leaveMessage()}
+      */
+     @NotNull
++    @Deprecated // Paper
+     public String getLeaveMessage() {
+-        return leaveMessage;
++        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().serialize(this.leaveMessage); // Paper
```

Comment:
All of these on the spot conversions are fine for an MVP, but from my experience running PGM with very hot rendering code this will affect server performance by a lot.

I suspect either Paper or Adventure will need to have a custom `TextComponent` that stores the legacy formatting in a field so de/serialization is `O(1)`.

Reply:
meh, most of the cases where stuff is converted, its not really hot code. as in, they are not called every tick. and ontop of that, these methods are deprecated anyways, if these turn out to be a performance bottleneck, we can just yell at ppl to update to the new methods 🤷

Reply:
> so de/serialization is `O(1)`.
Nitpicking, but _amortized_`O(1)`.